### PR TITLE
[ADD] Doc: information on which tests were not converted yet

### PR DESCRIPTION
**Pull request by Zinston**
When implementing the tests for the delete, backspace and enter keys,
certain tests from we3 had to be skipped. Please check these whenever a
new feature is implemented, and convert the tests that have become
relevant thanks to your feature, if any.

This is meant to be merged right after delete, backspace and enter tests.

### CODE REVIEW

File: packages/core/test/todo.md [hunk 1]
```diff
@@ -0,0 +1,2022 @@
...
+},
+{
+    name: "in p > b (ul > i before, i after): BACKSPACE at start",
+    content: '<ul><li><p><i>node to merge with</i></p></li></ul><p><b>◆node</b><i>to merge</i></p>',
+    steps: [{
+        key: 'BACKSPACE',
+    }],
+    test: '<ul><li><p><i>node to merge with</i><b>◆node</b><i>to merge</i></p></li></ul>',
+},
+```
```
Comment (dmo-odoo) [inline packages/core/test/todo.md]: Syntax highlighting is messed up in this part of the file. Is this a github issue or is there something wrong ?

File: packages/core/test/todo.md [hunk 2]
```diff
@@ -0,0 +1,2022 @@
...
+//     steps: [{
+//         start: "a:contents()[0]->0",
+//         end: "a:eq(1):contents()[0]->11",
+//         key: 'ENTER',
+//     }],
+//     test: {
+//         content: "<div class=\"unbreakable\"><a class=\"btn\" href=\"#\">Label</a><a class=\"btn\" href=\"#\">&#65279;ot to edit&#65279;</a></div>",
+//         start: "a:eq(1):contents()[0]->1",
+//     },
+// },
```
Comment (dmo-odoo) [inline packages/core/test/todo.md]: Why are these tests commented out, while the others are just "there" ?

File: packages/core/test/todo.md [hunk 3]
```diff
@@ -0,0 +1,2022 @@
...
+//     name: "in ul.list-group > li: ENTER at end",
+//     content: '<ul class="list-group"><li><p>dom to edit</p></li></ul>',
+//     steps: [{
+//         start: "p:contents()[0]->11",
+//         key: 'ENTER',
+//     }, {
+//         key: 'ENTER',
+//     }],
+//     test: '<ul class="list-group"><li><p>dom to edit</p></li><li><p><br/></p></li><li><p><br/>◆</p></li></ul>',
+// },
```
Comment (dmo-odoo) [inline packages/core/test/todo.md]: Same for every test that is purposefully commented out.

File: packages/core/test/todo.md (hunk 2 above)
Comment (Zinston) [reply packages/core/test/todo.md]: https://github.com/odoo-dev/jabberwock/pull/46/files#diff-0875bb23b2e3171746454f3566139fbaR1072

File: packages/core/test/todo.md [hunk 4]
```diff
@@ -0,0 +1,2024 @@
...
+    test: "<ul><li>▶<br/>◀</li></ul>",
+},
+{
+    name: "from p to li > p: DELETE on whole list",
+    content: "<p>dom not to edit▶</p><ul><li><p>dom to edit◀</p></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<p>dom not to edit◆</p>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 5]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in empty-li (no other li): DELETE -> 'a' (must write into it)",
+    content: "<ul><li><br/>◆</li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }, {
+        key: 'a',
+    }],
+    test: "<ul><li>a◆</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 6]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li (li after): DELETE at end (must move contents of second li to carret)",
+    content: "<ul><li>dom to&nbsp;◆</li><li>edit</li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li>dom to ◆edit</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 7]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li (li after): DELETE -> 'a' at end (must move contents of second li to carret, then write)",
+    content: "<ul><li>dom to&nbsp;◆</li><li>edit</li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }, {
+        key: 'a',
+    }],
+    test: "<ul><li>dom to a◆edit</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 8]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li > p: DELETE before single character",
+    content: "<ul><li><p>◆a</p></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li><p>▶<br/>◀</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 9]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li > p: DELETE -> 'a' before single character",
+    content: "<ul><li><p>◆a</p></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }, {
+        key: 'a',
+    }],
+    test: "<ul><li><p>a◆</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 10]
```diff
@@ -0,0 +1,2024 @@
...
+    test: "<ul><li><p>toto</p></li><li><p>xxx◆yyy</p></li><li><p>tutu</p></li></ul>",
+},
+{
+    name: "in li > p (b before - p after): DELETE at end",
+    content: "<ul><li><p>toto</p></li><li><p><b>x</b>xx◆</p><p><b>y</b>yy</p></li><li><p>tutu</p></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li><p>toto</p></li><li><p><b>x</b>xx◆<b>y</b>yy</p></li><li><p>tutu</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 11]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li > p (p.o_default_snippet_text after): DELETE at end",
+    content: '<ul><li><p>toto</p></li><li><p>xxx◆</p><p class="o_default_snippet_text">yyy</p></li><li><p>tutu</p></li></ul>',
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li><p>toto</p></li><li><p>xxx◆yyy</p></li><li><p>tutu</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "classes"

File: packages/core/test/todo.md [hunk 12]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in indented-ul > li: DELETE at end (must do nothing)",
+    content: "<ul><li><ul><li>dom to edit◆</li></ul></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: '<ul><li class="o_indent"><ul><li>dom to edit◆</li></ul></li></ul>',
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 13]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in indented-ul > li: DELETE at end -> 'a' at end (must write)",
+    content: "<ul><li><ul><li>dom to edit◆</li></ul></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }, {
+        key: 'a',
+    }],
+    test: '<ul><li class="o_indent"><ul><li>dom to edita◆</li></ul></li></ul>',
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 14]
```diff
@@ -0,0 +1,2024 @@
...
+    test: "<ul><li>d◆edit</li></ul>",
+},
+{
+    name: "in li: DELETE on selection of all contents",
+    content: "<ul><li>▶dom to edit◀</li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li>▶<br/>◀</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 15]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li: DELETE -> 'a' on selection of all contents",
+    content: "<ul><li>▶dom to edit◀</li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }, {
+        key: 'a',
+    }],
+    test: "<ul><li>a◆</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 16]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li: DELETE after character, before p",
+    content: "<ul><li>a</li><li>b◆<p>c</p></li></ul>",
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: "<ul><li>a</li><li><p>b◆c</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "irrelevant"

File: packages/core/test/todo.md [hunk 17]
```diff
@@ -0,0 +1,2024 @@
...
+    test: '<ul><li><p><i>node to merge</i> with◆<b>node</b> to merge</p></li></ul>',
+},
+{
+    name: "in li > p > i (b before - p > b after): DELETE at end",
+    content: '<ul><li><p><b>node to </b><i>merge with◆</i></p></li></ul><p><b>node</b> to merge</p>',
+    steps: [{
+        key: 'DELETE',
+    }],
+    test: '<ul><li><p><b>node to </b><i>merge with◆</i><b>node</b> to merge</p></li></ul>',
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 18]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li: ENTER -> 'a' on selection of all contents",
+    content: "<ul><li>▶dom to edit◀</li></ul>",
+    steps: [{
+        key: 'ENTER',
+    }, {
+        key: 'a',
+    }],
+    test: "<ul><li><br/></li><li>a◆</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "integration"

File: packages/core/test/todo.md [hunk 19]
```diff
@@ -0,0 +1,2024 @@
...
+//     name: "in ul.list-group > li: ENTER at end",
+//     content: '<ul class="list-group"><li><p>dom to edit</p></li></ul>',
+//     steps: [{
+//         start: "p:contents()[0]->11",
+//         key: 'ENTER',
+//     }, {
+//         key: 'ENTER',
+//     }],
+//     test: '<ul class="list-group"><li><p>dom to edit</p></li><li><p><br/></p></li><li><p><br/>◆</p></li></ul>',
+// },
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "classes"

File: packages/core/test/todo.md [hunk 20]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in indented-li with font: 2x ENTER at end",
+    content: '<ul><li><p>aaa</p></li><ul><li><p><font style="color: red;">dom to edit◆</font></p></li></ul><li><p>bbb</p></li></ul>',
+    steps: [{
+        key: 'ENTER',
+    }, {
+        key: 'ENTER',
+    }],
+    test: '<ul><li><p>aaa</p></li><li class="o_indent"><ul><li><p><font style="color:red">dom to edit</font></p></li></ul></li><li>▶<br/>◀</li><li><p>bbb</p></li></ul>',
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Move to "styles"

File: packages/core/test/todo.md [hunk 21]
```diff
@@ -0,0 +1,2024 @@
...
+    test: "<ul><li>d</li><li>◆edit</li></ul>",
+},
+{
+    name: "in li: ENTER on selection of all contents",
+    content: "<ul><li>▶dom to edit◀</li></ul>",
+    steps: [{
+        key: 'ENTER',
+    }],
+    test: "<ul><li><br/></li><li>▶<br/>◀</li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 22]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in li: 2x ENTER at end",
+    content: "<ul><li><p>dom to edit◆</p></li></ul>",
+    steps: [{
+        key: 'ENTER',
+    }, {
+        key: 'ENTER',
+    }],
+    test: "<ul><li><p>dom to edit◆</p></li></ul>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 23]
```diff
@@ -0,0 +1,2024 @@
...
+{
+    name: "in indented-li: 2x ENTER at end",
+    content: "<ul><li><p>aaa</p></li><li><ul><li><p>dom to edit◆</p></li></ul></li><li><p>bbb</p></li></ul>",
+    steps: [{
+        key: 'ENTER',
+    }, {
+        key: 'ENTER',
+    }],
+    test: '<ul><li><p>aaa</p></li><li class="o_indent"><ul><li><p>dom to edit</p></li></ul></li><li>▶<br/>◀</li><li><p>bbb</p></li></ul>',
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 24]
```diff
@@ -0,0 +1,2024 @@
...
+    test: "<p>▶<br/>◀</p>",
+},
+{
+    name: "in li > p > empty-b: ENTER",
+    content: "<ul><li><p><b><br/>◆</b></p></li></ul>",
+    steps: [{
+        key: 'ENTER',
+    }],
+    test: "<p><b>▶<br/>◀</b></p>",
+},
```
Comment (Zinston) [inline packages/core/test/todo.md]: Added in master-lists-age

File: packages/core/test/todo.md [hunk 25]
```diff
@@ -0,0 +1,2024 @@
...
+        7. [HR](#b-t-hr)
+        8. [Integration](#b-t-integration)
+3. [Enter (`insertParagraphBreak`)](#enter)
+    1. [To convert later](#e-todo)
+        1. [Classes](#e-t-classes)
+        2. [Buttons](#e-t-buttons)
+        3. [Lists](#e-t-lists)
+        4. [SPAN](#e-t-span)
+        5. [Integration](#e-t-integration)
+
```
Comment (Zinston) [inline packages/core/test/todo.md]: Add:
```markdown
4. [List (`toggleList`)](#list)
    1. [To convert later](#l-todo)
```

File: packages/core/test/todo.md [hunk 26]
```diff
@@ -0,0 +1,2024 @@
...
+//         },
+//         {
+//             key: 'a',
+//         }
+//     ],
+//     test: "<p><br/><br/><br/>a</p><div><a href=\"#\">dom to edit</a></div>",
+//         start: "p:contents()[3]->1",
+//     },
+// },
+```
```
Comment (Zinston) [inline packages/core/test/todo.md]: Add:
```markdown
## List (`toggleList`)<a name="list"></a>

### To convert later<a name="l-todo"></a>
(when their related feature is implemented)

#### Unbreakable<a name="l-t-unbreakable"></a>
```typescript
it('should turn a paragraph in a div and a paragraph into a list with two items', async () => {
    await testEditor({
        contentBefore: '<div><p>ab[cd</p></div><p>ef]gh</p>',
        stepFunction: toggleUnorderedList,
        contentAfter: '<div><ul><li>ab[cd</li></ul></div><ul><li>ef]gh</li></ul>',
    });
});
```
```

**Comment by Zinston** [comment]
@dmo-odoo Right after ListNode, I can adapt this, applying the above comments, then we should merge this no?

**Comment by Zinston** [comment]
@dmo-odoo Lists were merged so I adapted this PR. It can be merged now.

**Pull request merged**